### PR TITLE
feat(HIP-904) - TokenAirdropTransaction

### PR DESCRIPTION
**Pull request by 0xivanov**
**Description**:
- Implements `TokenAirdropTransaction` and it's unit/e2e tests
- Adds `PendingAirdropId` to the `TransactionRecord`

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-sdk-java/issues/1955 , https://github.com/hashgraph/hedera-sdk-java/issues/1953 , https://github.com/hashgraph/hedera-sdk-java/issues/1952

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: sdk/src/main/java/com/hedera/hashgraph/sdk/TokenAirdropTransaction.java [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+import java.util.Map;
+
+/**
+ * Token Airdrop
+ * An "airdrop" is a distribution of tokens from a funding account
+ * to one or more recipient accounts, ideally with no action required
+ * by the recipient account(s).
+ */
+public class TokenAirdropTransaction extends Transaction<TokenAirdropTransaction> {
+    // TODO extract common logic between transfer transaction
```
Comment (0xivanov) [inline sdk/src/main/java/com/hedera/hashgraph/sdk/TokenAirdropTransaction.java:24]: TODO

File: sdk/src/main/java/com/hedera/hashgraph/sdk/TokenTransferList.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
+// TODO add copyrights
```
Comment (0xivanov) [inline sdk/src/main/java/com/hedera/hashgraph/sdk/TokenTransferList.java:1]: TODO

File: sdk/src/main/java/com/hedera/hashgraph/sdk/TokenTransferList.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import com.google.protobuf.UInt32Value;
+import com.hedera.hashgraph.sdk.proto.AccountAmount;
+import com.hedera.hashgraph.sdk.proto.NftTransfer;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+public class TokenTransferList {
+        final TokenId tokenId;
```
Comment (0xivanov) [inline sdk/src/main/java/com/hedera/hashgraph/sdk/TokenTransferList.java:13]: formatting

File: sdk/src/main/java/com/hedera/hashgraph/sdk/PendingAirdropId.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        if (tokenId != null) {
+            builder.setFungibleTokenType(tokenId.toProtobuf());
+        } else if (nftId != null) {
+            builder.setNonFungibleToken(nftId.toProtobuf());
+        }
+        return builder.build();
+     }
+
+    @Override
+    public String toString() {
```
Comment (0xivanov) [inline sdk/src/main/java/com/hedera/hashgraph/sdk/PendingAirdropId.java:92]: package private

File: sdk/src/main/java/com/hedera/hashgraph/sdk/PendingAirdropRecord.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+public class PendingAirdropRecord {
+    private final PendingAirdropId pendingAirdropId;
+    private final long pendingAirdropAmount;
+
+    PendingAirdropRecord(PendingAirdropId pendingAirdropId, long pendingAirdropAmount) {
+        this.pendingAirdropId = pendingAirdropId;
+        this.pendingAirdropAmount = pendingAirdropAmount;
+    }
+
+    public com.hedera.hashgraph.sdk.proto.PendingAirdropRecord toProtobuf() {
```
Comment (0xivanov) [inline sdk/src/main/java/com/hedera/hashgraph/sdk/PendingAirdropRecord.java:34]: package private

File: sdk/src/main/java/com/hedera/hashgraph/sdk/AbstractTokenTransferLogic.java [hunk 6]
```diff
@@ -0,0 +1,367 @@
+package com.hedera.hashgraph.sdk;
```
Comment (0xivanov) [inline sdk/src/main/java/com/hedera/hashgraph/sdk/AbstractTokenTransferLogic.java:1]: license

File: sdk/src/testIntegration/java/com/hedera/hashgraph/sdk/test/integration/TokenAirdropTransactionIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+import org.junit.jupiter.api.Test;
+
+class TokenAirdropTransactionIntegrationTest {
+
+    private final int amount = 100;
+
+    @Test
+    @DisplayName("Transfers tokens when the account is associated")
+    void canAirdropAssociatedTokens() throws Exception {
+        var testEnv = new IntegrationTestEnv(1).useThrowawayAccount();
```
Comment (0xivanov) [inline sdk/src/testIntegration/java/com/hedera/hashgraph/sdk/test/integration/TokenAirdropTransactionIntegrationTest.java:59]: close the testEnv

File: sdk/src/main/java/com/hedera/hashgraph/sdk/PendingAirdropId.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+ */
+public class PendingAirdropId {
+    private final AccountId sender;
+    private final AccountId receiver;
+    @Nullable
+    private final TokenId tokenId;
+    @Nullable
+    private final NftId nftId;
+
+    PendingAirdropId(AccountId sender, AccountId receiver, TokenId tokenId) {
```
Comment (0xivanov) [inline sdk/src/main/java/com/hedera/hashgraph/sdk/PendingAirdropId.java:36]: make this public and add public setters

File: sdk/src/main/java/com/hedera/hashgraph/sdk/TokenAirdropTransaction.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+/**
+ * Token Airdrop
+ * An "airdrop" is a distribution of tokens from a funding account
+ * to one or more recipient accounts, ideally with no action required
+ * by the recipient account(s).
+ */
+public class TokenAirdropTransaction extends AbstractTokenTransferTransaction<TokenAirdropTransaction> {
+    /**
+     * Constructor.
+     */
```
Comment (agadzhalov) [inline sdk/src/main/java/com/hedera/hashgraph/sdk/TokenAirdropTransaction.java:40]: I think we don't need this comment here

**Review by agadzhalov — APPROVED** [review]
great work @0xivanov, LGTM 🚀

**Pull request merged**